### PR TITLE
[network] Integrate NetworkInterface paradigm with Healthchecker

### PR DESCRIPTION
**Pull request by gregnazario**
## Motivation

To fix some of the paradigms in the networking layer.  I introduce the `NetworkInterface`.  This is a trait that is implemented by every application to handle the internals of networking allowing for removal of duplication of logic, and abstracting away networking from actual applications.

This allows for:
* Getting `NetworkSender`s to multiple tasks to send messages
* Polling on `NetworkInterface` for new messages
* Storing application data that's specific to a per peer basis
* Hiding internals of networking from the application.

Additionally, I've created the `PeerMetadataStorage` to keep track of connections and their specifics for use by the `NetworkInterface`s.

I've integrated this with Healthchecker first because it's the simplest application protocol.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Unit tests for the Interface, and E2E tests elsewhere

## Related PRs

None

### CODE REVIEW

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs]: Yeah, I think the verdict is this will be removed, since we need multiple senders coming out of the interface.

File: network/src/application/db.rs
Comment (JoshLind) [reply network/src/application/db.rs:58]: Interesting.. My remaining concern here is how we're avoiding concurrency issues with hashmap entries? Or are we planning on not supporting concurrency? e.g., if two threads (or tasks), call this function in unison, requesting the same entry, they can both have access to it concurrently because the write lock is dropped before the entry is passed to `modifier`. I'm guessing this is desired behaviour? If so, why? Why not just grab the write lock and hold it until `modifier` completes? 

Note: I'm probably missing something :D

File: network/src/application/management.rs
Comment (JoshLind) [inline network/src/application/management.rs:13]: nit: "separate" -> "separately"?

File: network/src/protocols/health_checker/interface.rs
Comment (JoshLind) [inline network/src/protocols/health_checker/interface.rs:52]: Silly question: where do we set it to Disconnected? Or is it meant to go "Disconnecting" -> Removed entirely?

**Review by JoshLind — COMMENTED** [review]
Looks good, Greg! Some final questions (my only big one is about the race condition in `write`). But, other than that, it's cleaner and I'd be happy to approve (with my limited networking context :D)

File: network/src/application/interface.rs
Comment (davidiw) [reply network/src/application/interface.rs:18]: So it actually looks like we will still maintain independent DBs but have a consistent implementation?

**Review by davidiw — COMMENTED** [review]
For some reason, I thought that PeerDB would be a single object across the node instance with shared copies all over. I guess the goal here is instead to just have a consistent logical representation?

Overall seems good.

File: network/src/protocols/health_checker/mod.rs
Comment (davidiw) [inline network/src/protocols/health_checker/mod.rs:148]: It kinda feels like maybe we should make the PeerDB support a generic interface that contains Peer to PeerAppData so that we can merge this into that space as well -- otherwise we're still going to be maintaining this state everywhere and then we're trading a tens of lines of code for hundreds.

Comment (davidiw) [reply network/src/protocols/health_checker/mod.rs:148]: e.g., PeerDB<T> and support peer_data_read(&self, peer_id) -> &T and peer_data_write(&self, peer_id, value: T)

then we can specialize it in the local space

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs:18]: It wasn't read only, because originally it was writing application data and scores into the `PeerDb`, but that had a poor reception, so I nixed it.

The purpose of this function was also because I was originally going to make `PeerDb` a trait, but rust didn't like what I was doing.

File: network/src/protocols/health_checker/interface.rs
Comment (gregnazario) [reply network/src/protocols/health_checker/interface.rs:52]: Without a proper GC strategy, I'm just removing them entirely after going Disconnecting.

There's an option later to keep the data, and then remove it when it reaches a point, or when it's no longer in the trusted peer set.

**Comment by gregnazario** [comment]
@JoshLind 
> Interesting.. My remaining concern here is how we're avoiding concurrency issues with hashmap entries? Or are we planning on not supporting concurrency? e.g., if two threads (or tasks), call this function in unison, requesting the same entry, they can both have access to it concurrently because the write lock is dropped before the entry is passed to modifier. I'm guessing this is desired behaviour? If so, why? Why not just grab the write lock and hold it until modifier completes?

It's nested, so it grabs the lock, performs the change, and then released the lock at the end of `PeerDb.write()`.  I went a little fancy to try to remove the application developer's ability to hold onto a lock for too long / forget to handle it.

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs]: It's been cut.

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs]: It's been cut

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs:18]: > So it actually looks like we will still maintain independent DBs but have a consistent implementation?

@davidiw We have 1 DB, copying the `Arc` around of it.  This allows for the flexibility if we wanted to for some reason have multiple, but in reality they're all shared.

File: network/src/protocols/health_checker/mod.rs
Comment (gregnazario) [reply network/src/protocols/health_checker/mod.rs:148]: I got a little flak for centralizing this kind of data, but let me see what I can do.

File: network/src/application/management.rs
Comment (zekun000) [inline network/src/application/management.rs:15]: it seems every trait exposes the underlying db, are they all necessary? seems a leaky abstraction

**Review by zekun000 — COMMENTED** [review]
there's two abstraction from what I understand, one between application and peer manager another between peer manager and transport, but all three traits expose db seems weird to me.

File: network/src/application/management.rs
Comment (gregnazario) [reply network/src/application/management.rs:15]: I've cut some of these.  I was trying to not repeat code, while preventing certain implementations from changing state / disconnecting etc.  I agree it needed to be improved

**Review by JoshLind — COMMENTED** [review]
Thanks @gregnazario! The only real blocker from my side is just double checking that I'm not missing anything regarding the race condition.

Otherwise, my only high-level concerns are around the use of traits to avoid code duplication (I assume?). Sometimes just duplicating code is easier to understand for newbs like me than having many trait implementations and understanding what the interfaces do... e.g., exposing a `db()` method on each trait so that implementations can share database manipulation code trips me up a little bit. However, given that you've been blocked on this for a while, I don't want to block you further. Feel free to ping me about the race condition and I'll approve 😄

File: network/src/application/db.rs
Comment (gregnazario) [reply network/src/application/db.rs:63]: This file isn't supposed to exist, I literally already fixed this.

I've got to look and see what happened

File: network/src/application/interface.rs
Comment (gregnazario) [reply network/src/application/interface.rs:21]: It's to avoid code duplication, since this will be exactly the same code.  I'm basically trying to work around inheritance, which maybe I shouldn't be doing in rust.

**Review by JoshLind — APPROVED** [review]
Looks cleaner to me 😄 Just some nits (your call).

File: network/src/application/storage.rs
Comment (gregnazario) [reply network/src/application/storage.rs:40]: Yeah, I was trying to avoid this and the rust compiler didn't like me.  I'll take another pass on it.

File: network/src/application/storage.rs
Comment (gregnazario) [reply network/src/application/storage.rs:98]: Oh, I think this is because I was trying to generify it for multiple connections rather than just 1, but it doesn't make sense at this point.

A lot of the code in this networking depends on 1 peer to 1 connection, which causes a lot of confusion.

File: network/src/application/interface.rs
Comment (zekun000) [inline network/src/application/interface.rs:62]: PeerState seems an internal concept to network, why do we need to expose it to interface? it's confusing when I look at both PeerState and AppData

Comment (gregnazario) [reply network/src/application/interface.rs:62]: I had it on a separate interface, but I removed it since it was confusing.

The only reason this is on this interface is for the healthchecker, so maybe I should just move it there instead

File: network/src/protocols/health_checker/mod.rs
Comment (JoshLind) [inline network/src/protocols/health_checker/mod.rs:339]: nit: could use an `if let` to avoid the empty branch, no?

File: network/src/protocols/health_checker/mod.rs
Comment (JoshLind) [inline network/src/protocols/health_checker/mod.rs:382]: Same here: could use an `if let` to avoid the empty branch, no?

File: network/src/transport/mod.rs
Comment (JoshLind) [inline network/src/transport/mod.rs:150]: I don't know why we need this change 😆 I assume this doesn't affect anything (given that it's test only)?

**Review by JoshLind — APPROVED** [review]

File: network/src/transport/mod.rs
Comment (gregnazario) [reply network/src/transport/mod.rs:150]: This is a super annoying problem with the `ConnectionId`.

I had the option of adding another constructor, making the object public or crate public, and routing a bunch of stuff through so we can have `ConnectionId` comparison in tests.

Or, just increment it here, so we can have different `ConnectionId`s and not always have the same "default" mock one.

**Pull request closed without merging**
